### ISSUE TITLE
Advanced authentication methods

### ISSUE THREAD

**Issue by galah92** [bytebeamio/rumqtt#584]
I'd like to be able to approve/disapprove device connections based on advanced methods rather then simple password validation. For example, it would be great if we can have some type of a hook/middleware/notification saying a device is trying to connect with some credentials (client_id & password), and I'll be able to verify the password with my database. In the later phase, this will allow enforcing devices to send JWT-based password and have some kind of OAuth2-based authentication.

Is something like that currently supported?

Thank you

**Comment by henil** [comment]
We would like to have that feature, but its not something currently in priority.

**Comment by Nickztar** [comment]
This would be a great feature and is something that some other brokers support. Something like how HiveMQ does it: https://www.hivemq.com/extension/file-rbac-extension/

**Comment by swanandx** [comment]
> While not a priority, are you amenable to a PR?

Yup! PRs are always welcome 💯

> Do you have any thoughts on the design?

This wasn't priority so didn't really give it much thought, but definitely would love to discuss and ideate on design!

**Comment by andrewbaxter** [comment]
I don't have any specific ideas about how this would integrate since I'm not familiar with the code base, but here was my general idea:
* Instead of a hash map, config takes an object implementing an interface
* Create a static auth implementation, that gets a hash map
* Config file remains unchanged, but instantiates the static auth implementation
* When authentication is successful, the implementation returns a new object to be associated with the connection
* When receiving messages to a topic on the connection, use the associated object to optionally reject the message
* When about to notify on a subscribed connection, use the associated object to possibly not notify

My only idea on the specifics is that the config would take a `Box<dyn Authenticator>` for the authenticator.  I think the serialized config file struct and the config used by the broker objects themselves will need to be separated (so the broker doesn't store the deserialized config object).

**Comment by andrewbaxter** [comment]
I think splitting the serial structures from the programmatic config structure make sense also by decoupling serialization considerations from server requirements.  I already see there's a couple places where the config layout was dictated by serializer restrictions (i.e. `_` in names).

That said, is there a reason to repeat the authorizer configuration in multiple places in the config?  For example currently v4 and v5 have their own username/password maps.  If I move it to the top level I can get away with just duplicating the top level config structure for now.

**Comment by andrewbaxter** [comment]
Nevermind, I think I can leave auth in the nested structs without copying too much right now.

**Comment by swanandx** [comment]
Hey @galah92 , this issue would be solved with PR #761 !

you can define your fn like:
```rust
fn verify_from_db(client_id: String, username: String, password: String) -> bool {
  // do the verification and return true if verified, else false
}
```

and then pass it to `external_auth` field of `ConnectionSettings`.

for e.g. if you want it for `v4.1` server, you can do something like

```rust
// plz do it better way lmao
configs
    .v4
    .as_mut()
    .unwrap()
    .get_mut("1")
    .unwrap()
    .connections
    .external_auth = Some(Arc::new(verify_from_db));
```

**Comment by swanandx** [comment]
will be closing this issue!

note: we should add an example as well, please feel free to open PR to add one :)
